aioesphomeapi/client.py

```diff
@@ -1238,3 +1245,81 @@ def api_version(self) -> Optional[APIVersion]:
         if self._connection is None:
             return None
         return self._connection.api_version
+
+    async def subscribe_voice_assistant(
+        self,
+        handle_start: Callable[[], Coroutine[Any, Any, Optional[int]]],
+        handle_stop: Callable[[], Coroutine[Any, Any, None]],
+    ) -> Callable[[], None]:
+        """Subscribes to voice assistant messages from the device.
+
+        handle_start: called when the devices requests a server to send audio data to.
+                      This callback is asyncronous and returns the port number the server is started on.
+
+        handle_stop: called when the device has stopped sending audio data and the pipeline should be closed.
+
+        Returns a callback to unsubscribe.
+        """
+        self._check_authenticated()
+
+        t: Optional[asyncio.Task[Optional[int]]] = None
+
+        def _started(fut: asyncio.Task[Optional[int]]) -> None:
+            if self._connection is not None and not fut.cancelled():
+                port = fut.result()
+                if port is not None:
+                    self._connection.send_message(VoiceAssistantResponse(port=port))
+                else:
+                    _LOGGER.error("Server could not be started")
+                    self._connection.send_message(VoiceAssistantResponse(error=True))
+
+        def on_msg(msg: VoiceAssistantRequest) -> None:
+            command = VoiceAssistantCommand.from_pb(msg)
+            loop = asyncio.get_running_loop()
+            if command.start:
+                t = loop.create_task(handle_start())
+                t.add_done_callback(_started)
+            else:
+                loop.create_task(handle_stop())
```

Comment:
We should hold a reference to these so they do not get garbage collected mid execution

https://docs.python.org/3/library/asyncio-task.html#creating-tasks

From the asyncio tasks section:
> Important Save a reference to the result of this function, to avoid a task disappearing mid-execution. The event loop only keeps weak references to tasks. A task that isn’t referenced elsewhere may get garbage collected at any time, even before it’s done. For reliable “fire-and-forget” background tasks, gather them in a collection: